### PR TITLE
[SPARK-28940][SQL] Subquery reuse across all subquery levels

### PR DESCRIPTION
**Pull request by peter-toth**
### What changes were proposed in this pull request?

This PR enables subquery reuse across all subquery levels.
Please note that this feature works for adaptive query execution since https://github.com/apache/spark/pull/25471 but this PR:
- enables reuse for static execution (including Dynamic Partition Pruning) as well
- refactors AQE subquery reuse code a bit

Example query:
```
SELECT (SELECT avg(key) FROM testData), (SELECT (SELECT avg(key) FROM testData))
FROM testData
LIMIT 1
```
Plan before this PR:
```
CollectLimit 1
+- *(1) Project [Subquery scalar-subquery#268, [id=#231] AS scalarsubquery()#276, Subquery scalar-subquery#270, [id=#266] AS scalarsubquery()#277]
   :  :- Subquery scalar-subquery#268, [id=#231]
   :  :  +- *(2) HashAggregate(keys=[], functions=[avg(cast(key#13 as bigint))], output=[avg(key)#272])
   :  :     +- Exchange SinglePartition, true, [id=#227]
   :  :        +- *(1) HashAggregate(keys=[], functions=[partial_avg(cast(key#13 as bigint))], output=[sum#282, count#283L])
   :  :           +- *(1) SerializeFromObject [knownnotnull(assertnotnull(input[0, org.apache.spark.sql.test.SQLTestData$TestData, true])).key AS key#13]
   :  :              +- Scan[obj#12]
   :  +- Subquery scalar-subquery#270, [id=#266]
   :     +- *(1) Project [Subquery scalar-subquery#269, [id=#263] AS scalarsubquery()#275]
   :        :  +- Subquery scalar-subquery#269, [id=#263]
   :        :     +- *(2) HashAggregate(keys=[], functions=[avg(cast(key#13 as bigint))], output=[avg(key)#274])
   :        :        +- Exchange SinglePartition, true, [id=#259]
   :        :           +- *(1) HashAggregate(keys=[], functions=[partial_avg(cast(key#13 as bigint))], output=[sum#286, count#287L])
   :        :              +- *(1) SerializeFromObject [knownnotnull(assertnotnull(input[0, org.apache.spark.sql.test.SQLTestData$TestData, true])).key AS key#13]
   :        :                 +- Scan[obj#12]
   :        +- *(1) Scan OneRowRelation[]
   +- *(1) SerializeFromObject
      +- Scan[obj#12]
```
Plan after this PR:
```
CollectLimit 1
+- *(1) Project [ReusedSubquery Subquery scalar-subquery#241, [id=#148] AS scalarsubquery()#248, Subquery scalar-subquery#242, [id=#164] AS scalarsubquery()#249]
   :  :- ReusedSubquery Subquery scalar-subquery#241, [id=#148]
   :  +- Subquery scalar-subquery#242, [id=#164]
   :     +- *(1) Project [Subquery scalar-subquery#241, [id=#148] AS scalarsubquery()#247]
   :        :  +- Subquery scalar-subquery#241, [id=#148]
   :        :     +- *(2) HashAggregate(keys=[], functions=[avg(cast(key#13 as bigint))], output=[avg(key)#246])
   :        :        +- Exchange SinglePartition, true, [id=#144]
   :        :           +- *(1) HashAggregate(keys=[], functions=[partial_avg(cast(key#13 as bigint))], output=[sum#258, count#259L])
   :        :              +- *(1) SerializeFromObject [knownnotnull(assertnotnull(input[0, org.apache.spark.sql.test.SQLTestData$TestData, true])).key AS key#13]
   :        :                 +- Scan[obj#12]
   :        +- *(1) Scan OneRowRelation[]
   +- *(1) SerializeFromObject
      +- Scan[obj#12]
```

### Why are the changes needed?

Performance improvement.

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

Added new UTs.

### CODE REVIEW

**Comment by peter-toth** [comment]
cc @maryannxue @cloud-fan

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/PlanAdaptiveSubqueries.scala [hunk 1]
```diff
@@ -29,8 +29,6 @@ case class PlanAdaptiveSubqueries(
     plan.transformAllExpressions {
       case expressions.ScalarSubquery(_, _, exprId) =>
         subqueryMap(exprId.id)
-      case expressions.InSubquery(_, ListQuery(_, _, exprId, _)) =>
```
Comment (peter-toth) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/PlanAdaptiveSubqueries.scala]: @maryannxue I don't know why this case was required so I removed it. Please correct me if I was wrong.

**Comment by AngersZhuuuu** [comment]
I am doing a plugin for Spark to cache subquery data, hit cached data and replace LogicalPlan, support storage in memory, Alluxio.  some  code do same things. In spark's own  CacheManager,  it don't handle situation like we query hive data, then cache query data, but if cached query's source data has changed, It won't get new data when cache matched.

**Comment by peter-toth** [comment]
> I am doing a plugin for Spark to cache subquery data, hit cached data and replace LogicalPlan, support storage in memory, Alluxio. some code do same things. In spark's own CacheManager, it don't handle situation like we query hive data, then cache query data, but if cached query's source data has changed, It won't get new data when cache matched.

This PR has nothing to do with caching of data, it just enables reuse of subquery results (across all levels) by building up a map of plans. I might use wrong/confusing terminology by calling `subqueryCache` some of my vars. But let's see others' opinion on this.

**Comment by peter-toth** [comment]
retest this please

File: sql/core/src/test/scala/org/apache/spark/sql/DynamicPartitionPruningSuite.scala [hunk 2]
```diff
@@ -1238,7 +1238,10 @@ class DynamicPartitionPruningSuite
 
       val plan = df.queryExecution.executedPlan
       val countSubqueryBroadcasts =
-        plan.collectInPlanAndSubqueries({ case _: SubqueryBroadcastExec => 1 }).sum
+        plan.collectInPlanAndSubqueries({
```
Comment (maryannxue) [inline sql/core/src/test/scala/org/apache/spark/sql/DynamicPartitionPruningSuite.scala]: Nit: you could also do
```
      val countSubqueryBroadcasts =
        plan.collectInPlanAndSubqueries({ case _: SubqueryBroadcastExec => 1 }).sum
      val countReusedSubqueryBroadcasts =
        plan.collectInPlanAndSubqueries({
          case ReusedSubqueryExec(_: SubqueryBroadcastExec) => 1
        }).sum

      assert(countSubqueryBroadcasts == 1)
      assert(countReusedSubqueryBroadcasts == 1)
```

Comment (peter-toth) [reply sql/core/src/test/scala/org/apache/spark/sql/DynamicPartitionPruningSuite.scala]: ok, I changed

Comment (maryannxue) [reply sql/core/src/test/scala/org/apache/spark/sql/DynamicPartitionPruningSuite.scala]: We have `InSubqueryExec` now in DPP, so better keep it.

Comment (peter-toth) [reply sql/core/src/test/scala/org/apache/spark/sql/DynamicPartitionPruningSuite.scala]: Ok, I've reverted the change.

**Comment by peter-toth** [comment]
cc @mgaido91, your comments are welcome too

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/InsertAdaptiveSparkPlan.scala [hunk 3]
```diff
@@ -41,9 +41,6 @@ case class InsertAdaptiveSparkPlan(
 
   private val conf = session.sessionState.conf
 
-  // Subquery-reuse is shared across the entire query.
-  private val subqueryCache = new TrieMap[SparkPlan, BaseSubqueryExec]()
```
Comment (mgaido91) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/InsertAdaptiveSparkPlan.scala]: we can ermove the import right?

**Review by mgaido91 — COMMENTED** [review]
the change seems reasonable to me. I don't like too much adding a new map to the QueryExecution but I have no better suggestion at the moment...

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/InsertAdaptiveSparkPlan.scala (hunk 3 above)
Comment (peter-toth) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/InsertAdaptiveSparkPlan.scala]: No, unfortunately we can't, `stageCache` still needs it.

**Comment by peter-toth** [comment]
After https://github.com/apache/spark/pull/26015 some of the new UTs showed and error and I realized that the new feature of that PR (exchange reuse across all subquery levels) conflicts with my implementation of subquery reuse across all subquery levels.
This was because the `subqueryCache` that I added to the `QueryExecution` contained a cached subquery that was later removed by the new version of `ReuseExchange` rule. The subquery that was cached had a parent `Exchange` node that was replaced to a `ReusedExchangeExec` node.
In the last commit I changed my subquery reuse implementation to something similar to exchange reuse implementation.
I believe this new approach is a bit less efficient but the advantage is that there is no more need for `subqueryCache` in `QueryExecution` and overall the required change is smaller than before.

**Comment by peter-toth** [comment]
retest this please

**Pull request closed without merging**